server/src/main/java/org/elasticsearch/shutdown/PluginShutdownService.java

```diff
@@ -0,0 +1,108 @@
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * The {@link PluginShutdownService} is used for the node shutdown infrastructure to signal to
+ * plugins that a shutdown is occurring, and to check whether it is safe to shut down.
+ */
+public class PluginShutdownService implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(PluginShutdownService.class);
+    public List<ShutdownAwarePlugin> plugins;
+
+    public PluginShutdownService(@Nullable List<ShutdownAwarePlugin> plugins) {
+        this.plugins = plugins == null ? Collections.emptyList() : plugins;
+    }
+
+    /**
+     * Return all nodes shutting down from the given cluster state
+     */
+    public static Set<String> shutdownNodes(final ClusterState clusterState) {
+        return NodesShutdownMetadata.getShutdowns(clusterState)
+            .map(NodesShutdownMetadata::getAllNodeMetadataMap)
+            .map(Map::keySet)
+            .orElse(Collections.emptySet());
+    }
+
+    /**
+     * Return all nodes shutting down with the given shutdown type from the given cluster state
+     */
+    public static Set<String> shutdownTypeNodes(final ClusterState clusterState, final SingleNodeShutdownMetadata.Type shutdownType) {
+        return NodesShutdownMetadata.getShutdowns(clusterState)
+            .map(NodesShutdownMetadata::getAllNodeMetadataMap)
+            .map(m -> m.entrySet().stream()
+                .filter(e -> e.getValue().getType() == shutdownType)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)))
+            .map(Map::keySet)
+            .orElse(Collections.emptySet());
+    }
+
+    /**
+     * Check with registered plugins whether the shutdown is safe for the given node id and type
+     */
+    public boolean readyToShutdown(String nodeId, SingleNodeShutdownMetadata.Type shutdownType) {
+        // TODO: consider adding debugging information (a message about why not?)
+        // TODO: consider adding more fine-grained status rather than true/false
+        for (ShutdownAwarePlugin plugin : plugins) {
+            try {
+                if (plugin.safeToShutdown(nodeId, shutdownType) == false) {
+                    logger.trace("shutdown aware plugin [{}] is not yet ready for shutdown", plugin);
+                    return false;
+                }
+            } catch (Exception e) {
+                logger.warn("uncaught exception when retrieving whether plugin is ready for node shutdown", e);
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Signal to plugins the nodes that are currently shutting down
+     */
+    public void signalShutdown(final ClusterState state) {
+        Set<String> shutdownNodes = shutdownNodes(state);
+        for (ShutdownAwarePlugin plugin : plugins) {
+            try {
+                plugin.signalShutdown(shutdownNodes);
```

Comment:
Similarly, it would be nice if `state` was passed as an extra argument here, so that plugins that don't currently have their own reference to `ClusterService` can look at the current cluster state.